lib/src/git.rs

```diff
@@ -1319,16 +1322,45 @@ fn push_refs(
             GitPushError::InternalGitError(err)
         }
     })?;
-    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs.iter().copied().collect();
+    let mut remaining_remote_refs_expected_locations =
+        qualified_remote_refs_expected_locations.clone();
     let mut push_options = git2::PushOptions::new();
     let mut proxy_options = git2::ProxyOptions::new();
     proxy_options.auto();
     push_options.proxy_options(proxy_options);
     let mut callbacks = callbacks.into_git();
+    callbacks.push_negotiation(|updates| {
+        for update in updates {
+            let dst_refname = update
+                .dst_refname()
+                .expect("Expect reference name to be valid UTF-8");
+            let expected_location = (*qualified_remote_refs_expected_locations
+                .get(dst_refname)
+                .expect("Push is trying to move a ref it wasn't asked to move"))
+            .clone();
```

Comment:
nit: `.clone()` isn't needed (and `Option<&CommitId>` is easier to deal with than `&Option<CommitId>`.)